fuel/src/main/kotlin/com/github/kittinunf/fuel/core/extensions/Authentication.kt

```diff
@@ -0,0 +1,39 @@
+package com.github.kittinunf.fuel.core.extensions
+
+import com.github.kittinunf.fuel.core.Headers
+import com.github.kittinunf.fuel.core.Request
+import com.github.kittinunf.fuel.util.encodeBase64ToString
+
+class AuthenticatedRequest(private val wrapped: Request) : Request by wrapped {
```

Comment:
[Q] Reading up until this left me with a bit of a question in my head, it looks to me that we have multiple request type that is separated in a different class because they are fundamentally different, right?

We have
* DownloadRequest.kt
* SuspendableRequest.kt
* CancellableRequest.kt
etc.

They are significantly implemented the Request interface in the different way I guess.

What are roughly the criteria that you use to break this down into a separated class like this one? 

In other words, I think this should just be a feature inside the Request class which should turn into just an extension over Request?

Like why didn't you do this instead? So, we don't actually need the wrapper? (basically like Formatter.kt)

``` Kotlin
private fun Request.basic(username: String, password: String): Request 

private fun Request.bearer(token: String): Request 

// public api
fun Request.authenticate(username: String, password: String) 
```

Reply:
>  because they are fundamentally different, right?
That, but also their behaviour is both isolated but also dependent on the other classes/instances/functions in those files.

> They are significantly implemented the Request interface in the different way I guess.
Correct, but note that none of them are mutually exclusive. They are not necessarily overwriting the implementation of Request, but all wrapping, so you keep the old behaviour AND get the new one on top. A DefaultRequest can become Cancellable, Downloading a file, Suspendable, and Uploading data, all at the same time. 

> What are roughly the criteria that you use to break this down into a separated class like this one?
- If the feature can be isolated completely [so that you have an a-cyclic directed graph)
- [soft criteria] Is not dependent on other features, [so that you have an a-cyclic directed graph]
- Is not going to be called _by default_ (for example parameters could be a feature, but it's nice to have on `Request`) [so that you don't need to import it always]
- You probably call the new methods it makes available right after the extension method (for example, you can actually do `.download().file().upload().blob().download().name()`, but you probably will do `.download().file().name().upload().blob()` ). [so it doesn't look disgusting]

> In other words, I think this should just be a feature inside the Request class which should turn into just an extension over Request?
This is actually now possible! yes, `.download()` could be an extension function on `Request` and then everything still works. The only reason I did NOT do that with download and upload yet is simply because I moved them first, and `Request` was not yet an interface, and therefore it would have broken a lot of code.

In other words: There is NO strong reason for `download` and `upload` to still be on `Request` interface, except for not needing to import it right away.

> Like why didn't you do this instead? So, we don't actually need the wrapper? (basically like Formatter.kt)
I had different ideas for authentication until it boiled down to this. The main reason is that I did not want to clutter the namespace with all different types of authentication (I can think of 8 of the top of my head) just like .upload() is called and .blob etc don't clutter the default namespace. Since I know we will probably support at least 5 of the different authentications, some of them requiring even more parameters and methods, I thought it might have been nice to extract to a wrapper right away.

Reply:
Ah, thanks for detailed explanation. I believe that you have tried others ways already before it boils down to this. I guess it might be hard to see the value as we don't have all of the authentications yet but I think it is good as where it stands now.

Reply:
Easy to roll back if we hate it!, then authentication can just return self.